### PR TITLE
responsive compiler prototype: frontend queries and tests

### PR DESCRIPTION
**Pull request by mppf**
This PR builds upon previous PRs to add frontend queries for parsing and
some C++ tests of the new compiler code. This PR completes the series of
PRs #17585, #17586, #17587, #17588, #17589.

Split out from PR #17583, which is a start at the compiler revamp effort
described in the
[1.24 release notes ongoing efforts](https://chapel-lang.org/releaseNotes/1.24/04-ongoing.pdf).

This PR adds Frontend queries and C++ tests of several of the features
added to the compiler in the previous PRs. The Frontend queries implement
parsing and one that (as a demonstration) lists the defined top-level
names in each module parsed.

Closes #17583.

 - [x] full local testing
 
Reviewed by @e-kayrakli  and @dlongnecke-cray - thanks!

### CODE REVIEW

File: compiler/next/include/chpl/Frontend/FrontendQueries.h [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+#include "chpl/AST/Location.h"
+#include "chpl/AST/ModuleDecl.h"
+#include "chpl/Queries/Context.h"
+
+#include <vector>
+
+namespace chpl {
+
+// This could alternatively be a class with static methods
+namespace FrontendQueries {
```
Comment (e-kayrakli) [inline compiler/next/include/chpl/Frontend/FrontendQueries.h:36]: should we try to stick to snake_case for namespace names? Or at least that's what `chpl` and `ast` suggests to me.

File: compiler/next/test/Frontend/testFrontendQueries.cpp [hunk 2]
```diff
@@ -0,0 +1,406 @@
+/*
```
Comment (e-kayrakli) [inline compiler/next/test/Frontend/testFrontendQueries.cpp:1]: Reading this test was fun!

**Review by e-kayrakli — APPROVED** [review]
Looks pretty straightforward and good! The tests here are very useful to get a better context of where we are in this effort IMO.

The only thing that stood out for me is the namespace naming convention (or lack thereof?). I prefer if we have it more consistent from the beginning, if possible.

File: compiler/next/include/chpl/Frontend/FrontendQueries.h (hunk 1 above)
Comment (mppf) [reply compiler/next/include/chpl/Frontend/FrontendQueries.h:36]: That's a reasonable suggestion. I've renamed it to `namespace frontend` on my branch with many other renamings that will form a subsequent PR.

**Review by dlongnecke-cray — APPROVED** [review]

File: compiler/next/include/chpl/Frontend/FrontendQueries.h (hunk 1 above)
Comment (mppf) [reply compiler/next/include/chpl/Frontend/FrontendQueries.h:36]: PR #17609 will get to it

**Pull request merged**